onnxruntime/test/providers/provider_test_utils.cc

```diff
@@ -68,6 +68,42 @@ void Check(const OpTester::Data& expected_data, const Tensor& output_tensor,
   }
 }
 
+template <>
+void Check<uint8_t>(const OpTester::Data& expected_data,
+                    const Tensor& output_tensor,
+                    const std::string& provider_type) {
+  auto& expected_tensor = expected_data.data_.Get<Tensor>();
+  auto* expected = expected_tensor.template Data<uint8_t>();
+  auto* output = output_tensor.template Data<uint8_t>();
+  auto size = output_tensor.Shape().Size();
+
+  if (expected_data.sort_output_) {
+    // if order can be jumbled in the output of an operator, sort both the
+    // expected and output buffers prior to
+    // comparison this is a "best-effort" algo and should satisfy the
+    // requirement for the few ops that do require this
+    // support without investing in a more sophisticated infrastructure for the
+    // same
+    sort_expected_and_actual_buffers<uint8_t>(expected, output, size);
+  }
+
+  // NNAPI's rounding is a bit different than CPU provider
```

Comment:
I am not so sure about this part, since some of the output from NNAPI are within +/-1 of the result of CPU provider
This may be caused by different rounding mode of NNAPI (I think NNAPI use HALF_AWAY_FROM_ZERO)

Reply:
I'd suggest checking with Yufeng whether this is an expected/acceptable diff in general. If it's NNAPI specific, and we are sure that's due to a rounding diff, we need to make the threshold configurable.

---
In reply to: [544740088](https://github.com/microsoft/onnxruntime/pull/6155#discussion_r544740088) [](ancestors = 544740088)

Reply:
Sync'ed with Yufeng offline, updated comments